File: src/main/java/com/dambra/paul/stringcalculator/calculationParts/Operation.java
package com.dambra.paul.stringcalculator.calculationParts;

public class Operation extends CalculationPart {
    private final String operator;
    private final double nextValue;

    public Operation(String operator, String nextValue) {
        this.operator = operator;
        this.nextValue = Double.valueOf(nextValue);
    }

    @Override
    public double Calculate(double total) {
        return next.Calculate(performCalculation(total));
    }

    private double performCalculation(double total) {
        switch (operator) {
            case "+":
                total += nextValue;
                break;
            case "-":
                total -= nextValue;
                break;
            case "/":
                total /= nextValue;
                break;
            default:
                total *= nextValue;
                break;
        }
        return total;
    }
}

File: src/main/java/com/dambra/paul/stringcalculator/calculationParts/Operation.kt
package com.dambra.paul.stringcalculator.calculationParts

class Operation(private val operator: String, nextValue: String) : CalculationPart() {
    private val nextValue: Double = java.lang.Double.valueOf(nextValue)

    override fun calculate(total: Double) =
            next.calculate(performCalculation(total))

    private fun performCalculation(total: Double) =
            when (operator) {
                "+" -> total + nextValue
                "-" -> total - nextValue
                "/" -> total / nextValue
                else -> total * nextValue
            }
}
